### ISSUE TITLE
AWS complaining to be unable to load credentials

### ISSUE THREAD

**Issue by spheenik** [cantaloupe-project/cantaloupe#620]
I just set up my first instance of Cantaloupe ever, so it might be my fault (using the 5.0.5 release build).
The images are on a minio-Instance I also host.

When cantaloupe tries to get a source image, there is a relatively long delay, and then the following exception is logged in the output:

```
23:19:13.032 [qtp1130894323-28] DEBUG e.i.l.c.c.S3Cache - newDerivativeImageInputStream(): bucket: cantaloupe-cache; key: image/50104aea1b2e2e28425929141d26b462/1df1efc163e914e8eb8d42f829af8633.jpg
2023-02-02T23:19:13.298807212Z 23:19:13.298 [qtp1130894323-28] DEBUG e.i.l.c.c.FilesystemCache - getSourceImageFile(): /var/cache/cantaloupe/source/50/10/4a/50104aea1b2e2e28425929141d26b462 
2023-02-02T23:19:13.301174482Z 23:19:13.301 [qtp1130894323-28] DEBUG s.a.a.r.p.AwsRegionProviderChain - Unable to load region from software.amazon.awssdk.regions.providers.SystemSettingsRegionProvider@618e8023:Unable to load region from system settings. Region must be specified either via environment variable (AWS_REGION) or  system property (aws.region).
2023-02-02T23:19:13.301855494Z 23:19:13.301 [qtp1130894323-28] DEBUG s.a.a.r.p.AwsRegionProviderChain - Unable to load region from software.amazon.awssdk.regions.providers.AwsProfileRegionProvider@6e27ee00:No region provided in profile: default
2023-02-02T23:19:21.063417718Z 23:19:21.063 [qtp1130894323-28] DEBUG s.a.a.r.p.AwsRegionProviderChain - Unable to load region from software.amazon.awssdk.regions.providers.InstanceProfileRegionProvider@7f49d1c9:Unable to contact EC2 metadata service.
2023-02-02T23:19:21.072215517Z 23:19:21.071 [qtp1130894323-28] DEBUG s.a.a.a.c.AwsCredentialsProviderChain - Unable to load credentials from SystemPropertyCredentialsProvider(): Unable to load credentials from system settings. Access key must be specified either via environment variable (AWS_ACCESS_KEY_ID) or system property (aws.accessKeyId).
2023-02-02T23:19:21.072252348Z software.amazon.awssdk.core.exception.SdkClientException: Unable to load credentials from system settings. Access key must be specified either via environment variable (AWS_ACCESS_KEY_ID) or system property (aws.accessKeyId).
2023-02-02T23:19:21.072261468Z 	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:98)
2023-02-02T23:19:21.072273668Z 	at software.amazon.awssdk.auth.credentials.internal.SystemSettingsCredentialsProvider.resolveCredentials(SystemSettingsCredentialsProvider.java:58)
2023-02-02T23:19:21.072280628Z 	at software.amazon.awssdk.auth.credentials.AwsCredentialsProviderChain.resolveCredentials(AwsCredentialsProviderChain.java:91)
2023-02-02T23:19:21.072293928Z 	at software.amazon.awssdk.awscore.client.handler.AwsClientHandlerUtils.createExecutionContext(AwsClientHandlerUtils.java:76)
2023-02-02T23:19:21.072301408Z 	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.createExecutionContext(AwsSyncClientHandler.java:68)
2023-02-02T23:19:21.072306829Z 	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.lambda$execute$1(BaseSyncClientHandler.java:97)
2023-02-02T23:19:21.072318319Z 	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.measureApiCallSuccess(BaseSyncClientHandler.java:167)
2023-02-02T23:19:21.072323249Z 	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:94)
2023-02-02T23:19:21.072332029Z 	at software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:45)
2023-02-02T23:19:21.072338109Z 	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:55)
2023-02-02T23:19:21.072343019Z 	at software.amazon.awssdk.services.s3.DefaultS3Client.headObject(DefaultS3Client.java:5012)
2023-02-02T23:19:21.072351529Z 	at edu.illinois.library.cantaloupe.source.S3Source.getObjectAttributes(S3Source.java:295)
2023-02-02T23:19:21.072357559Z 	at edu.illinois.library.cantaloupe.source.S3Source.checkAccess(S3Source.java:279)
2023-02-02T23:19:21.072362539Z 	at edu.illinois.library.cantaloupe.resource.ImageRequestHandler.handle(ImageRequestHandler.java:338)
2023-02-02T23:19:21.072381630Z 	at edu.illinois.library.cantaloupe.resource.iiif.v3.ImageResource.doGET(ImageResource.java:130)
2023-02-02T23:19:21.072388660Z 	at edu.illinois.library.cantaloupe.resource.HandlerServlet.handle(HandlerServlet.java:97)
2023-02-02T23:19:21.072393610Z 	at edu.illinois.library.cantaloupe.resource.HandlerServlet.doGet(HandlerServlet.java:35)
2023-02-02T23:19:21.072420260Z 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
2023-02-02T23:19:21.072437801Z 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
2023-02-02T23:19:21.072444241Z 	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:791)
2023-02-02T23:19:21.072463221Z 	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550)
2023-02-02T23:19:21.072470011Z 	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
2023-02-02T23:19:21.072477381Z 	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434)
2023-02-02T23:19:21.072497202Z 	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
2023-02-02T23:19:21.072516992Z 	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
2023-02-02T23:19:21.072538862Z 	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
2023-02-02T23:19:21.072544443Z 	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349)
2023-02-02T23:19:21.072554993Z 	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
2023-02-02T23:19:21.072568353Z 	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
2023-02-02T23:19:21.072577433Z 	at org.eclipse.jetty.server.Server.handle(Server.java:516)
2023-02-02T23:19:21.072586843Z 	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
2023-02-02T23:19:21.072599593Z 	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
2023-02-02T23:19:21.072605124Z 	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
2023-02-02T23:19:21.072617344Z 	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
2023-02-02T23:19:21.072633024Z 	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
2023-02-02T23:19:21.072646244Z 	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
2023-02-02T23:19:21.072659834Z 	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
2023-02-02T23:19:21.072682125Z 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
2023-02-02T23:19:21.072694285Z 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
2023-02-02T23:19:21.072704915Z 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
2023-02-02T23:19:21.072727116Z 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:135)
2023-02-02T23:19:21.072738196Z 	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:773)
2023-02-02T23:19:21.072743956Z 	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:905)
2023-02-02T23:19:21.072768096Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2023-02-02T23:19:21.072782567Z 
```
After that, it is able to get the source image from minio and process it correctly.

I have configured the credentials according to the documentation:
```
S3Source.access_key_id = *access key*
S3Source.secret_key = *secret*
```

Did I do something wrong? Do I really have to pass the credentials via the environment?

**Comment by spheenik** [comment]
To add, it also complains about an unset region, do I need to set it?

```
23:19:13.301 [qtp1130894323-28] DEBUG s.a.a.r.p.AwsRegionProviderChain - Unable to load region from software.amazon.awssdk.regions.providers.SystemSettingsRegionProvider@618e8023:Unable to load region from system settings. Region must be specified either via environment variable (AWS_REGION) or  system property (aws.region).
2023-02-02T23:19:13.301855494Z 23:19:13.301 [qtp1130894323-28] DEBUG s.a.a.r.p.AwsRegionProviderChain - Unable to load region from software.amazon.awssdk.regions.providers.AwsProfileRegionProvider@6e27ee00:No region provided in profile: default
2023-02-02T23:19:21.063417718Z 23:19:21.063 [qtp1130894323-28] DEBUG s.a.a.r.p.AwsRegionProviderChain - Unable to load region from software.amazon.awssdk.regions.providers.InstanceProfileRegionProvider@7f49d1c9:Unable to contact EC2 metadata service.
2023-02-02T23:19:21.072215517Z 2
```

**Comment by DiegoPino** [comment]
Hi, you are missing a config key, e.g

```
# !! AWS region. Only needed for AWS endpoints.
S3Source.region = us-east-1
```

**Comment by spheenik** [comment]
Hi Diego,
I figured that my endpoint is not AWS (I am using Minio), so I did not need to set it, since it says it's only needed for AWS.
And even if setting `S3Source.region` in the configuration file, I get the message.

**Comment by DiegoPino** [comment]
@spheenik but does S3 work? If so you can dismiss the message (change the logging level). There are a few messages like the credential chain loader from the environment (then ones AWS provides) that are not really errors but feel (and read like that)

**Comment by spheenik** [comment]
Yes, it works! But, iirc, if I did not supply the region, there was a looong delay (5 secs or so), until it continued working.
Anyway, I supplied a JavaDelegate, which reads the properties from the environment on startup, and puts them into the environment variables where the AWS-SDK expects them. Noise-free debug-log since then.

Btw.: It would be nice to allow the JavaDelegate to use the Contaloupe Configuration. I used a hack to extract it, by supplying a ConfigurationExtractor class in the correct package:

```java
package edu.illinois.library.cantaloupe.config;

public class ConfigurationExtractor {

    public static Configuration getInstance() {
        return ConfigurationFactory.getInstance();
    }

}
```

Anyway, all is fine. Issue can be closed.
